dvc/external_repo.py

```diff
@@ -65,66 +68,206 @@ def clean_repos():
+    def get_checksum(self, path):
+        raise NotImplementedError
+
+    def check_recursive_imports(self, path):
+        """Raise RecursiveImportError if path_info contains recursively added
+        DVC outs.
+        """
+        path_info = PathInfo(self.root_dir) / path
+        self._recursive_outputs(path_info)
+
+    def _recursive_outputs(self, path_info, recursive=False):
+        # if path_info is a non-dvc directory, we need to check for
+        # recursively added dvc files
+        fetch_infos = []
+        for root, dirs, files in self.repo_tree.walk(path_info):
+            root_path = PathInfo(root)
+            for name in dirs + files:
+                if self.repo_tree.isdvc(root_path / name):
+                    if recursive:
+                        fetch_infos.append(self._fetch_info(root_path / name))
+                    else:
+                        raise RecursiveImportError(
+                            path_info.relative_to(self.root_dir)
+                        )
+        return fetch_infos
+
+    def _fetch_info(self, path_info):
+        if self.repo_tree.isdvc(path_info):
+            (out,) = self.find_outs_by_path(path_info, strict=False)
+            filter_info = path_info.relative_to(out.path_info)
+        else:
+            filter_info = None
+        return path_info, filter_info
+
+    def get_external(self, path, to_info, recursive=False, **kwargs):
         """
         Pull the corresponding file or directory specified by `path` and
-        checkout it into `to_info`.
+        into `to_info`.
 
         It works with files tracked by Git and DVC, and also local files
         outside the repository.
         """
-        out = None
         path_info = PathInfo(self.root_dir) / path
 
-        with suppress(OutputNotFoundError):
-            (out,) = self.find_outs_by_path(fspath(path_info), strict=False)
-
-        try:
-            if out and out.use_cache:
-                self._pull_cached(out, path_info, to_info)
-                return
+        if not self.repo_tree.exists(path_info):
+            raise PathMissingError(path, self.url)
 
-            # Check if it is handled by Git (it can't have an absolute path)
-            if os.path.isabs(path):
-                raise FileNotFoundError
+        fetch_infos = [self._fetch_info(path_info)]
+        if self.repo_tree.isdir(path_info) and not self.repo_tree.isdvc(
+            path_info
+        ):
+            fetch_infos.extend(self._recursive_outputs(path_info, recursive))
 
-            fs_copy(fspath(path_info), fspath(to_info))
-        except FileNotFoundError:
-            raise PathMissingError(path, self.url)
+        self._fetch_external(fetch_infos, **kwargs)
 
-    def _pull_cached(self, out, path_info, dest):
         with self.state:
-            tmp = PathInfo(tmp_fname(dest))
-            src = tmp / path_info.relative_to(out.path_info)
+            self.repo_tree.copytree(path_info, to_info)
```

Comment:
`get_external` is now essentially just a `fetch` and then `checkout` (checkout is done for dvc outs in `RepoTree.copytree`)